lucene/core/src/java/org/apache/lucene/search/similarities/BM25Similarity.java

```diff
@@ -221,8 +251,8 @@ public final SimScorer scorer(float boost, CollectionStatistics collectionStats,
 
     @Override
     public float score(float freq, long encodedNorm) {
-      double norm = cache[((byte) encodedNorm) & 0xFF];
-      return weight * (float) (freq / (freq + norm));
+      float norm = cache[((byte) encodedNorm) & 0xFF];
+      return weight * tf(freq, norm);
```

Comment:
the crazy casting is not an optimization but instead to ensure that scores dont go backwards with increasing tf, decreasing norm, etc. it has to behave "semimonotonic" (like described in java Math api docs) so that wand type optimizations are correct. so if you changed the api to be all doubles, it would mean we'd have to use BigDecimal which would br slower.

Reply:
Thanks, very good point.